ethereum/blockwatch/block_watcher_test.go

```diff
@@ -67,25 +72,398 @@ func TestWatcher(t *testing.T) {
+	BlockNumber: uint64(30),
+	TxHash:      common.HexToHash("0xd9bb5f9e888ee6f74bedcda811c2461230f247c205849d6f83cb6c3925e54586"),
+	TxIndex:     uint(0),
+	BlockHash:   common.HexToHash("0x6bbf9b6e836207ab25379c20e517a89090cbbaf8877746f6ed7fb6820770816b"),
+	Index:       uint(0),
+	Removed:     false,
+}
+
+var errUnexpected = errors.New("Something unexpected")
+
+type filterLogsRecusivelyTestCase struct {
+	Label                     string
+	rangeToFilterLogsResponse map[string]filterLogsResponse
+	Err                       error
+	Logs                      []types.Log
+}
+
+func TestFilterLogsRecursively(t *testing.T) {
+	from := int64(10)
+	to := int64(20)
+	testCases := []filterLogsRecusivelyTestCase{
+		filterLogsRecusivelyTestCase{
+			Label: "HAPPY_PATH",
+			rangeToFilterLogsResponse: map[string]filterLogsResponse{
+				"10-20": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+			},
+			Logs: []types.Log{logStub},
+		},
+		filterLogsRecusivelyTestCase{
+			Label: "TOO_MANY_RESULTS_INFURA_ERROR",
+			rangeToFilterLogsResponse: map[string]filterLogsResponse{
+				"10-20": filterLogsResponse{
+					Err: errors.New(infuraTooManyResultsErrMsg),
+				},
+				"10-15": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				"16-20": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+			},
+			Logs: []types.Log{logStub, logStub},
+		},
+		filterLogsRecusivelyTestCase{
+			Label: "TOO_MANY_RESULTS_INFURA_ERROR_DEEPER_RECURSION",
+			rangeToFilterLogsResponse: map[string]filterLogsResponse{
+				"10-20": filterLogsResponse{
+					Err: errors.New(infuraTooManyResultsErrMsg),
+				},
+				"10-15": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				"16-20": filterLogsResponse{
+					Err: errors.New(infuraTooManyResultsErrMsg),
+				},
+				"16-18": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				"19-20": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+			},
+			Logs: []types.Log{logStub, logStub, logStub},
+		},
+		filterLogsRecusivelyTestCase{
+			Label: "TOO_MANY_RESULTS_INFURA_ERROR_DEEPER_RECURSION_FAILURE",
+			rangeToFilterLogsResponse: map[string]filterLogsResponse{
+				"10-20": filterLogsResponse{
+					Err: errors.New(infuraTooManyResultsErrMsg),
+				},
+				"10-15": filterLogsResponse{
+					Logs: []types.Log{
+						logStub,
+					},
+				},
+				"16-20": filterLogsResponse{
+					Err: errUnexpected,
+				},
+			},
+			Err: errUnexpected,
+		},
+		filterLogsRecusivelyTestCase{
+			Label: "UNEXPECTED_ERROR",
+			rangeToFilterLogsResponse: map[string]filterLogsResponse{
+				"10-20": filterLogsResponse{
+					Err: errUnexpected,
+				},
+			},
+			Err: errUnexpected,
+		},
+	}
+
+	meshDB, err := meshdb.NewMeshDB("/tmp/leveldb_testing/" + uuid.New().String())
+	require.NoError(t, err)
```

Comment:
This is far from the only place this happens, but we should remember to call `Close` when instantiating a new `MeshDB` in our tests. I created https://github.com/0xProject/0x-mesh/issues/209 to address this.